src/GraphQL/Server.hs

```diff
@@ -208,22 +209,31 @@ instance forall ks t f m. (MonadThrow m, KnownSymbol ks, BuildFieldResolver m f,
     NamedFieldExecutor "" (queryError ("buildFieldResolver got non AST.Field" <> show f <> ", query probably not normalized"))
 
 
--- TODO we can probably use closed type families for RunFieldsType and
--- FieldHandler for better error reporting in case the user uses some
--- unexpected type.
+-- We only allow Field and Argument :> Field combinations:
+type family RunFieldsType (m :: Type -> Type) (a :: [Type]) = (r :: Type) where
+  RunFieldsType m '[Field ks t] = Field ks t
+  RunFieldsType m '[a :> b] = a :> b
+  RunFieldsType m ((Field ks t) ': rest) = Field ks t :<> RunFieldsType m rest
+  RunFieldsType m ((a :> b) ': rest) = (a :> b) :<> RunFieldsType m rest
+  RunFieldsType m a = TypeLits.TypeError (
+    'TypeLits.Text "All field entries in an Object must be Field or Argument :> Field. Got: " 'TypeLits.:<>: 'TypeLits.ShowType a)
+
 
 class RunFields m a where
-  type RunFieldsType m a :: Type
+  type RunFieldsHandler m a :: Type
   -- Runfield is run on a single QueryTerm so it can only ever return
```

Comment:
Please get case of `Runfield` correct (I think it's supposed to be `runFields` but am not sure).